exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/xsort/managed/ExternalSortBatch.java

```diff
@@ -0,0 +1,1456 @@
+    // by closing the resultsIterator after all results are returned downstream.
+
+    MergeSort memoryMerge = new MergeSort(context, allocator, opCodeGen);
+    try {
+      sv4 = memoryMerge.merge(bufferedBatches, this, container);
+      if (sv4 == null) {
+        sortState = SortState.DONE;
+        return IterOutcome.STOP;
+      } else {
+        logger.info("Completed in-memory sort. Memory = {}",
+                allocator.getAllocatedMemory());
+        resultsIterator = memoryMerge;
+        memoryMerge = null;
+        sortState = SortState.DELIVER;
+        return IterOutcome.OK_NEW_SCHEMA;
+      }
+    } finally {
+      if (memoryMerge != null) {
+        memoryMerge.close();
+      }
+    }
+  }
+
+  /**
+   * Perform merging of (typically spilled) batches. First consolidates batches
+   * as needed, then performs a final merge that is read one batch at a time
+   * to deliver batches to the downstream operator.
+   *
+   * @return always returns OK_NEW_SCHEMA
+   */
+
+  private IterOutcome mergeSpilledRuns() {
+    logger.info("Starting consolidate phase. Batches = {}, Records = {}, Memory = {}, In-memory batches {}, spilled runs {}",
+                inputBatchCount, inputRecordCount, allocator.getAllocatedMemory(),
+                bufferedBatches.size(), spilledRuns.size());
+
+    // Consolidate batches to a number that can be merged in
+    // a single last pass.
+
+    int mergeCount = 0;
+    while (consolidateBatches()) {
+      mergeCount++;
+    }
+    stats.addLongStat(Metric.MERGE_COUNT, mergeCount);
+
+    // Merge in-memory batches and spilled runs for the final merge.
+
+    List<BatchGroup> allBatches = new LinkedList<>();
+    allBatches.addAll(bufferedBatches);
+    bufferedBatches.clear();
+    allBatches.addAll(spilledRuns);
+    spilledRuns.clear();
+
+    logger.info("Starting merge phase. Runs = {}, Alloc. memory = {}", allBatches.size(), allocator.getAllocatedMemory());
+
+    // Do the final merge as a results iterator.
+
+    CopierHolder.BatchMerger merger = copierHolder.startFinalMerge(schema, allBatches, container, mergeBatchRowCount);
+    merger.next();
+    resultsIterator = merger;
+    sortState = SortState.DELIVER;
+    return IterOutcome.OK_NEW_SCHEMA;
+  }
+
+  private boolean consolidateBatches() {
+
+    // Determine additional memory needed to hold one batch from each
+    // spilled run.
+
+    int inMemCount = bufferedBatches.size();
+    int spilledRunsCount = spilledRuns.size();
+
+    // Can't merge more than will fit into memory at one time.
+
+    int maxMergeWidth = (int) (mergeMemoryPool / targetMergeBatchSize);
+    maxMergeWidth = Math.min(mergeLimit, maxMergeWidth);
+
+    // If we can't fit all batches in memory, must spill any in-memory
+    // batches to make room for multiple spill-merge-spill cycles.
+
+    if (inMemCount > 0) {
+      if (spilledRunsCount > maxMergeWidth) {
+        spillFromMemory();
```

Comment:
The function is obligated to spill. If it can't (disk full? file system errors? OOM?), it will throw an exception. So, there is no status to return. Even if we did return a status, each caller would then be obligated to handle the error case, and so on up the call chain. Since exceptions do that for us, we use them rather than returning an error status.